package/MDAnalysis/core/groups.py

```diff
@@ -1849,17 +1953,21 @@ def unwrap(self, compound='fragments', reference='com', inplace=True):
         #  unidimensionally whereas the general multi-compound case involves
         #  more indexing and is therefore slower. Leaving separate for now.
         if comp == 'group':
-            positions = mdamath.make_whole(unique_atoms, inplace=False)
+            positions = unique_atoms.positions
+            spread = positions.ptp(axis=0)
+            spread = distances.transform_RtoS(spread, self.dimensions)
+            if np.any(spread > .5):
```

Comment:
Im trying to remember if this is still true for triclinic boxes or if you need a trig correction...

Reply:
I think this is fine as is, since you already do the trig transform on the axis system. I don't have any formal proof, though (one may exist in some textbook, I guess?). I just couldn't come up with any breaking examples. (Note that this assumes that bonds cannot be larger than half the box vectors, which should be mostly ok).

Reply:
So I've been trying to prove this a bit more formally for triclinic cells.
The current ptp->RtoS strategy essentially means

 _Take each compound's bounding box diagonal vector, decompose it into the triclinic vectors, and see if any is resulting component is larger than half the corresponding box vector._

So the question here is: are the projected diagonals always at least as large as they can be in every component using this method? I can already find cases where they aren't, so I'll dig some more. I have a hunch that I should be able to just compare the untransformed ptp difference to the orthorhombic vectors.

A direct alternative is to RtoS transform the entire system prior to unwrapping, but that seems costly...

Reply:
If you write it down in equations, I am happy to put on my "reviewer 2" hat and look over it ;-).

Reply:
Thanks, @orbeckst, I'm definitely taking you up on that offer! Please review the reasoning here while I clean up code/testing for pushing.

After much head scratching, and some performance measurements, my conclusions are as follows regarding the shortcuts in non-rectangular boxes:
_Here I'm defining as 'safe' either a compound that is wholly within half of each vector and therefore assumed not to need unwrapping, or the box volume defined by the half box vectors_
- Transforming the diagonals of compound bounding boxes to S space may wrongly decide a compound is safe when it isn't. The most obvious counter-example is to imagine
  1. the safe parallelepiped defined by half of each box vector (in essence a scaled down version of the box, corresponding to an eighth of its volume), which limits the safe space a compound can occupy, and
  1. the smallest rectangular parallelepiped that circumscribes the one in i). This rectangular parallelepiped, which is equivalent to its own bounding box, will have one diagonal that is entirely contained in the safe volume. Yet, parts of the parallelepiped will be outside, and it will span more than half of one or more box vectors.
- The lazy way out is to transform all atoms (not just diagonal vectors) to S space and compare compound spans there against 0.5. In my tests this incurs in a penalty between 15 and 20% of performance. Let's try to avoid that.

## Strategy
The idea is to find, for each compound, the parallelepiped with same angles as the box and varying $l_a$, $l_b$ and $l_c$ so that it circumbscribes the compound's rectangular bounding box. We then compare those lengths to half the box vector lengths to decide whether a compound is safe.

### Inscribed rectangular parallelepipeds
The ASCII figures below represents top views of box parallelepipeds, down along the z-axis. In each, the two parallelograms represent the bottom ( $z=0$ ) and top ( $z=c_z$ ) faces. The inscribed rectangular parallelepiped, with faces parallel to the $xy$, $xz$ and $yz$ planes, is the one that fits inside the projected intersection of the $z=0$ and $z=c_z$ parallelograms. Its face on the $xy$ plane is marked by points $\mathbf{r_1},\mathbf{r_2},\mathbf{r_4},\mathbf{r_3}$. The origin is marked by ```#```. Depending on the $c$ vector, there are four possible cases:

Case I: $c_y \ge 0$  and  $\frac{c_x}{c_y} > \frac{b_x}{b_y}$
```
               _________________
              /z=cz            /
       __l___/__________      /
      /     /|r3   r4| /     /
     /     /_|_______|/_____/
    /        r1      /r2
   #_z=0____________/
```

Case II: $c_y \ge 0$ and $\frac{c_x}{c_y} \le \frac{b_x}{b_y}$
```
      _________________
     /z=cz            / 
    /  ______________/__
   /  /|r4      r3| /  /
  /__/_|__________|/  /
    /  r1        r2  /
   #_z=0____________/
```

Case III: $c_y < 0$  and  $\frac{c_x}{c_y} > \frac{b_x}{b_y}$  (equivalent to case I, but with symmetric $c_x$, $c_y$)
```
            _________________
           /z=0             /
    __l___/__________      /
   /     /|r3   r4| /     /
  /     #_|_______|/_____/
 /        r1      /r2
/_z=cz___________/
```

Case IV: $c_y < 0$  and $\frac{c_x}{c_y} \le \frac{b_x}{b_y}$ (equivalent to case II, but with symmetric $c_x$, $c_y$)
```
       _________________
      /z=0             /
     /  ______________/__
    /  /|r4      r3| /  /
   #__/_|__________|/  /
     /  r1        r2  /
    /_z=cz___________/
```

Note that, at least regarding inscribed volumes, cases I and III are equivalent because they correspond to parallelepipeds symmetrized through the z-axis (conversion of one into the other requires inverting the signs of $c_x$ and $c_y$). Likewise for cases II and IV. We shall look only at cases I and II (later, we'll see how to interconvert between cases I and II, which will simplify the algortithm since we'll only have to deal with one case).

For case I the coordinates of $\mathbf{r_1}$, $\mathbf{r_2}$, $\mathbf{r_3}$ and $\mathbf{r_4}$ (all in the same $xy$ plane) are given by:
$\mathbf{r_1} = (b_x + l, c_y)$
$\mathbf{r_2} = (a_x + c_x - l, c_y)$
$\mathbf{r_3} = (b_x + l, b_y)$
$\mathbf{r_4} = (a_x + c_x - l, b_y)$
where $a_x$, $b_x$, etc. are the respective components of the triclinic volume vectors, and $l$ is the length of the segment indicated in the figure:
$l = c_x - c_y \frac{b_x}{b_y}$

$e_x$ and $e_y$ are the edges of this inscribed rectangle along $x$ and $y$, resp., with lengths:
$e_x = a_x + c_x - b_x - 2l = a_x - c_x - b_x + 2c_y\frac{b_x}{b_y}$
$e_y = b_y - c_y$

For case II (simpler, because the inscribed rectangular parallelipiped is now defined by parallelogram vertices, rather than by their intersection) $\mathbf{r_1}$, $\mathbf{r_2}$, $\mathbf{r_3}$ and $\mathbf{r_4}$ are:
$\mathbf{r_1} = (b_x, c_y)$
$\mathbf{r_2} = (a_x + c_x, c_y)$
$\mathbf{r_3} = (b_x, b_y)$
$\mathbf{r_4} = (a_x + c_x, b_y)$

and the $e_x$ and $e_y$ side lengths:
$e_x = a_x + c_x - b_x$
$e_y = b_y - c_y$

### Circumscribed parallelepiped
The above reasoning can be reversed to find the triclinc parallelepiped that circumscribes a given rectangular one (in our case, the latter will be the compounds' bounding boxes). We'll want the circumscribing parallelepiped to have the same angles as the box, for direct comparison with the reference safe volume. We find it by finding scaling factors $f_a$, $f_b$ and $f_c$ that, when multiplied by the box vectors, yield an inscribed rectangular parallelepiped equal to our bounding box.

If we have a rectangular bounding box that we want circumscribed, with edge lengths $r_x$, $r_y$ and $r_z$, the equations above become:

_for case I_
$f_a a_x - f_b b_x + f_c (2c_y\frac{b_x}{b_y} - c_x) = r_x$ (note the cancelling out of $f_b$ in the $\frac{b_x}{b_y}$ ratio)

_for case II_
$f_a a_x - f_b b_x + f_c c_x = r_x$

The y and z dimensions are the same in both cases
$f_b b_y - f_c c_y = r_y$
$f_c c_z = r_z$

In matrix form:
_case I_

$$\begin{bmatrix}
 a_x & -b_x & 2c_y\frac{b_x}{b_y} - c_x \\
 0   & b_y  & -c_y  \\
 0   &      &  c_z
\end{bmatrix} \cdot
\begin{bmatrix}
f_a \\
f_b \\
f_c
\end{bmatrix} =
\begin{bmatrix}
r_x \\
r_y \\
r_z
\end{bmatrix}$$


_case II_

$$\begin{bmatrix}
 a_x & -b_x &  c_x  \\
 0   & b_y  & -c_y  \\
 0   &      &  c_z
\end{bmatrix} \cdot
\begin{bmatrix}
f_a \\
f_b \\
f_c
\end{bmatrix} =
\begin{bmatrix}
r_x \\
r_y \\
r_z
\end{bmatrix}$$

With $\mathbf{P}$ the coefficients' matrix, $\mathbf{f}$ the factors' vector and $\mathbf{r}$ the rectangular parallelepiped's dimensions vector (the latter corresponding to the all-positive bounding box diagonal vector):
$\mathbf{P} \cdot \mathbf{f} = \mathbf{r}$
$\mathbf{f} = \mathbf{P}^{-1} \cdot \mathbf{r}$

After obtaining $\mathbf{f}$, _from only the bounding box's diagonal_ deciding which molecules are safe is a direct comparison with 0.5.

### Case I and II simplification
We can also see by comparing $\mathbf{P}$ for either case, which differ only in their topmost-rightmost element, that conversion from case I to II is simply a matter of subtracting $2l$ from $c_x$, where $l = c_x - b_x c_y / b_y$ (see the case I figure above). A geometric visualization of this equivalence is to picture the case I above, but with the $z=0$ parallelogram shifted by $l$ to the right, and the $z=c_z$ parallelogram shifted by $l$ to the left. The inscribed rectangular parallelogram will be exactly the same, but now circumscribed by a parallelogram of case II.

For code simplification, all box types will be converted to case II before solving for $\mathbf{f}$. Note that this conversion and the determination of $\mathbf{P}$ is rather cheap, and need only be done once per frame.

### Negative $b_x$ and other oddities
The above cases cover all possible box types with $a$ and $b$ vectors in the $xy$ plane, and positive $a_x$, $b_y$, and $c_z$ (note that the [common assumption](https://manual.gromacs.org/current/reference-manual/algorithms/periodic-boundary-conditions.html) that $|b_x| \leq \frac{a_x}{2}$, $|c_x| \leq \frac{a_x}{2}$ and $|c_y| \leq \frac{b_y}{2}$ needed not be invoked).

There remains the (rare?) case when $b_x < 0$, corresponding to boxes with $\angle ab > 90^{\circ}$. It can be easily seen (sorry, tired of ASCII drawings) that mirroring of $b$ and $c$ through the $yz$ plane yields equivalent (rotated) parallelograms of one of the cases above.

Because of all the symmetry/rotation equivalences it seems that this method would be valid for _any_ boxes that have $a$ and $b$ in the $xy$ plane (proof left as an exercise to the reader), but the ones discussed until here should be the only valid MD ones.

## Further optimization
There may still be molecules within the safe volume that are not identified as such with the above method (it's fine to unwrap them even if unneeded; we just lose a bit of performance). We can directly exclude those with bounding box diagonals outside the rectangular parallelepiped that circumscribes the safe volume. The remaining, which by now should be a small fraction, can be refined by RtoS transformation of all atoms, although the bookkeeping cost may offset the possible gains.

## Alternative approaches
I also looked into an apparently simpler approach: the problem with the RtoS transformation of the box diagonal is that the result might not be the S-space diagonal with largest spread. Initially, I was RtoS transforming only the all-positive diagonal of a compound's bounding box; yet there are three other bounding-box diagonals that will transform to vectors in S-space with different $x$ and $y$ spans ( the span in $z$ is always the same due to $a$ and $b$ lying in the $xy$ plane).

This points to a conceptually simpler solution than the lengthy one above: to find if any of the four diagonals of a molecule's bounding box has a span greater than 0.5 in S-space. I tried finding if this could be simplified as finding a specific offending diagonal among the four, but couldn't come up with anything simpler than just RtoS transforming all four diagonals.

Now, RtoS transforming all four diagonals seems easy, but we can see it's one more transformation than transforming all atoms of a three-particle molecule abundantly used in simulation. So I left it at the solution above, which does only one matrix multiplication operation per molecule. Perhaps someone else wants to take a gander at this approach, which would simplify code somewhat.

## Conclusions
In the end I was able to get back the 10-20% lost to RtoS transforming the entire system (yay!)

The code to do this isn't terribly complicated. I implemented it as a simple function that checks box type, converts to case II, and returns $\mathbf{P}^{-1}$ for multiplication later. It would be important, however, to somehow link all this explanation to the code. Do you think a link to a comment in a PR-review in github is solid enough?

Regarding testing, I'm thinking of synthetically creating all four cases (eight cases, with negative $b_x$), but maybe you have other sugestions?

Cheers!

Reply:
I'll have some time Friday to look at it. Thanks for the ASCII art and using math notation.